lib/internal/modules/esm/resolve.js

```diff
@@ -43,6 +47,28 @@ const {
   ERR_UNSUPPORTED_ESM_URL_SCHEME,
 } = require('internal/errors').codes;
 
+const DEFAULT_CONDITIONS = ObjectFreeze(['node', 'import']);
+const DEFAULT_CONDITIONS_SET = new Set(['node', 'import']);
+
+function getConditionsSet(conditions) {
+  if (conditions !== DEFAULT_CONDITIONS && ArrayIsArray(conditions)) {
+    if (ArrayPrototypeIncludes(conditions, 'require')) {
+      throw new ERR_INVALID_ARG_VALUE('conditions', conditions,
+                                      'may not contain "require"');
+    }
+    if (!ArrayPrototypeIncludes(conditions, 'import')) {
+      throw new ERR_INVALID_ARG_VALUE('conditions', conditions,
+                                      'has to contain "import"');
+    }
+    if (!ArrayPrototypeIncludes(conditions, 'node')) {
+      throw new ERR_INVALID_ARG_VALUE('conditions', conditions,
+                                      'has to contain "node"');
+    }
+    return new Set(conditions);
+  }
+  return DEFAULT_CONDITIONS_SET;
+}
```

Comment:
I would really prefer here that we pick one of the following two options:

a) Allow all conditions to be changed without these errors, then carefully document how conditions need to be passed through via arguments with a clear warning there about what happens if it is done wrong.
b) Do not allow these conditions to be disabled at all, and make the condition list purely additive. In that case these errors can still apply, just as all being exclusions.

I don't mind which of the above we choose - but I think it would be much clearer conceptually if we pick one path and stick with it.

Reply:
That said I would still prefer treating user conditions as an extra layer on top of the Node.js internal conditions that apply (no repetition necessary, so no errors above), but am happy to just see the errors disabled too with a careful docs comment.

Reply:
I think both a) and b) have the potential for future breakage if we add conditions to the default resolver, the same for any custom conditions we may pass in. I think no matter the way forward we need to admit that breakage and thus errors can occur. I like the current approach since it carves out a small subset of known desired conditions that we have talked about and are in resolution semantics already.

Reply:
@bmeck I don't think this argument holds for loaders since it is always superseded in effect by the base breaking scenario for any default resolver.

**Adding any new condition** to the default resolver is a breaking change, regardless of this loader hook. What really counts is how established user patterns are - kind of like how browsers could add `Array.prototype.includes`  as non-breaking but not `Array.prototype.contains`.

With the current approach, we risk new conditions actually throwing loaders that hard code conditions. Eg a loader that is experimenting with a "production" condition breaks when that condition is made to throw since it is a core condition. This seems detrimental to forward compatibility of new experimentation.

With both (a) and (b) a loader that experiments with new conditions then layering into that condition being brought into core continues to function fine.

Reply:
I think erring on the side of being permissive (and not forcing conditions) would be the correct direction. If we can agree on a direction here, happy to seal the deal and make the change.

My preferred solution would be "node accepts the conditions passed and will only implicitly add default". Testing tools like JSDOM have good reasons *not* to want the node-specific implicit conditions. I was fine adding the current exceptions on the premise that removing them in follow-up PRs is easier than adding them.

Reply:
Ok @jkrems if that is your preference I can get behind it with these errors removed but a very careful documentation comment about this, and specifically the need for loader to pass these conditions and risks if they don't.

Reply:
Removed the exception and added a sentence to the docs.

Reply:
Looks great, thanks.